# **Task 3: Fraud Detection System**

**Description:**

Develop a fraud detection system using a dataset like the Credit Card Fraud Dataset.

**Steps:**

**1. Data Preprocessing:**

○ Handle imbalanced data using techniques like SMOTE or undersampling.

**2. Model Training:**

○ Train a Random Forest or Gradient Boosting model to detect fraudulent
transactions.

**3. Model Evaluation:**

○ Evaluate the system's precision, recall, and F1-score.

**4. Testing Interface:**

○ Create a simple interface (e.g., a command-line input) to test the fraud detection
system.

**Outcome:**

● A Python script capable of detecting fraudulent transactions with evaluation metrics and
a testing interface.

# **Installation and Import Dataset and Libraries**

In [39]:
from google.colab import files
uploaded = files.upload()
import pandas as pd
df = pd.read_csv('creditcard.csv')


Saving creditcard.csv to creditcard (1).csv


In [40]:
df.head(8)
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

In [41]:
df.head()
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [42]:
!pip install -q imbalanced-learn

# import libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler

# **1. Data Preprocessing:**
○ Handle imbalanced data using techniques like SMOTE or undersampling.

In [43]:
X = df.drop(['Class'], axis=1)
y = df['Class']


In [44]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Handle Imbalance using SMOTE
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_scaled, y)


# **2. Model Training:**
○ Train a Random Forest or Gradient Boosting model to detect fraudulent
transactions.

In [45]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

In [46]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

# **3. Model Evaluation:**
○ Evaluate the system's precision, recall, and F1-score.

In [73]:
y_predic = model.predict(X_test)
print("\n Classification Report (Precision, Recall, F1-Score):\n ")
print(classification_report(y_test , y_predic))
print('\n')
print('Accuracy score for model : ',accuracy_score(y_test,y_predic))


 Classification Report (Precision, Recall, F1-Score):
 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56750
           1       1.00      1.00      1.00     56976

    accuracy                           1.00    113726
   macro avg       1.00      1.00      1.00    113726
weighted avg       1.00      1.00      1.00    113726



Accuracy score for model :  0.9998944832316269


# **4. Testing Interface:**
○ Create a simple interface (e.g., a command-line input) to test the fraud detection
system.

In [78]:
def test_transaction(data):
    if len(data) != 30:
        return

    data = np.array(data).reshape(1, -1)

    # Convert to Dataframe with correct column names for scaler
    data_df = pd.DataFrame(data, columns=X.columns)

    scaled_data = scaler.transform(data_df)

    # Predict using trained model
    prediction = model.predict(scaled_data)

    return " Fraudulent Transaction Detected!" if prediction[0] == 1 else " Legitimate Transaction."


In [79]:
import warnings
warnings.filterwarnings("ignore")


In [82]:
print("\n Running a sample test on a real transaction (row 0 of dataset)\n")
sample_input = X.iloc[0].values
result = test_transaction(sample_input)
print(f"Result: {result}")



 Running a sample test on a real transaction (row 0 of dataset)

Result:  Legitimate Transaction.


**Manual Testing**





In [ ]:
print("\n Enter 30 values (Time, V1-V28, Amount):")
user_input = [float(input(f"Enter feature {i+1}: ")) for i in range(30)]
print("Result:", test_transaction(user_input))


 Enter 30 values (Time, V1-V28, Amount):
Enter feature 1: 1000
Enter feature 2:  -1.359807
Enter feature 3: -0.072781


manually enter 30 values (features) related to a credit card transaction, including:

1. Time

2. V1 to V28 (anonymized features)

3. Amount

like:

Enter 30 values (Time, V1-V28, Amount):

Enter feature 1: 10000

Enter feature 2: -1.359807

Enter feature 3: -0.072781
...
to feature 30

The output either:

Result: Legitimate Transaction

or

Result: Fraudulent Transaction Detected!